two phases flow codes/Dambreak_LevelSet_script.py

```diff
@@ -0,0 +1,239 @@
+p = Function(ScalarSpace)
+p0 = Function(ScalarSpace)
+
+del_phi_c = TestFunction(ScalarSpace1) #level set function
+phi_c = Function(ScalarSpace1) #level set function
+phi_cr = Function(ScalarSpace1) #level set function after reinitialization
+phi_c0 = Function(ScalarSpace1)
+phi_cr0 = Function(ScalarSpace1) #level set function after reinitialization
+
+del_n_phi = TestFunction(VectorSpace1) 
+n_phi = Function(VectorSpace1) #normal of interface calculate from level set function
+
+V = FunctionSpace(mesh, 'P', 1)
+
+
+## Define initial conditions
+epsilon=0.5*pow(scale/res,0.9)#related to the size of element
+interface_x=0.3*scale
+interface_y=0.5*scale
+initial_phi_c = CharacterLevelSetFunction(param1=interface_x, param2=interface_y, param3=epsilon, element=V.ufl_element(),degree=1)#initial level set function
+initial_v = Expression(('0.', '0.'), degree=1)
+initial_p = Expression('(x[0] < x0 && x[1] < y0) ? (y0-x[1])*1.0 : 0.0',x0=interface_x,y0=interface_y, degree=1) #initial the pressure field of water 
+
+#%%
+## initialization
+
+# project(initial_phi_c, ScalarSpace1)
+# project(initial_v, VectorSpace)
+# project(initial_p, ScalarSpace)
+phi_c.assign(project(initial_phi_c, ScalarSpace1))
+phi_c0.assign(project(initial_phi_c, ScalarSpace1))
+u_c0.assign(project(initial_v, VectorSpace))
+p0.assign(project(initial_p*initial_phi_c, ScalarSpace)) 
+#mutiply the inital level-set value can make the initial calculation step more stable, but it has no effect on the result
+# can also use p0.assign(project(initial_p, ScalarSpace)) 
+
+
+# set parameters
+Re=2000.
+Fr=0.7
+rho0=1.
+nu0=1.
+rho1=0.001
+nu1=1.
+# write file
+file_pp = File(foldername+'/p_Re%d' %Re + '_Fr%.3s.pvd'%Fr)
+file_vv = File(foldername+'/v_Re%d' %Re + '_Fr%.3s.pvd'%Fr)
+file_c = File(foldername+'/c_Re%d' %Re + '_Fr%.3s.pvd'%Fr)
+file_c << (phi_c0,t)
+file_pp << (p0,t)
+
+file_p = XDMFFile(foldername+'/p_Re%d' %Re + '_Fr%.3s.xdmf'%Fr)
+file_v = XDMFFile(foldername+'/v_Re%d' %Re + '_Fr%.3s.xdmf'%Fr)
+file_phi = XDMFFile(foldername+'/phi_Re%d' %Re + '_Fr%.3s.xdmf'%Fr)
+#file_p.write_checkpoint(p, "p", t, append=False)
+#file_v.write_checkpoint(u_c, "v", t, append=False)
+#file_phi.write_checkpoint(phi_c, "phi", t, append=False)
+
+
+g_=-1.
+rou= phi_c*rho0+(1-phi_c)*rho1
+mu= phi_c*nu0+(1-phi_c)*nu1
+
+ndim = mesh.geometry().dim()
+
+# viscous operator
+def a(phi,chi, psi): return Constant(0.5) * inner(grad(phi) + grad(chi).T, grad(psi) + grad(psi).T)
+# divergence operator
+def b(phi, psi): return inner(div(phi), psi)
+# non-linear convection operator
+def c(phi, chi, psi): return dot(dot(grad(chi), phi), psi)
+# Crank-Nicholson schema average
+def d(phi, psi): return Constant(0.5) * (phi + psi)
+
+#n_phi=grad(phi_c)/pow(dot( grad(phi_c), grad(phi_c) ), 0.5)
+g = Expression(('0.','g'),degree=1,g=g_)
+
+# interface normal n = ▽ϕ/|ϕ|
+F_n_phi = (dot(grad(phi_c)/pow(dot( grad(phi_c), grad(phi_c) ), 0.5), del_n_phi) - dot(n_phi,del_n_phi)) * dx
```

Comment:
This seems to be a stationary equation. Why is this so? The other equations contain a time derivative.

Reply:
I still have questions here, if I use project to obtain this interface normal field:
```python
normalphi=grad(phi_c)/pow(dot( grad(phi_c), grad(phi_c) ), 0.5)
n_phi=project(normalphi,VectorSpace1)
```
which doesn't work
but solving this equation work well
```python
F_n_phi = (dot(grad(phi_c)/pow(dot( grad(phi_c), grad(phi_c) ), 0.5), del_n_phi) - dot(n_phi,del_n_phi)) * dx
```

Reply:
I cannot tell either in detail without an error message but the project method should work if the space is compatible to the projected expression.

Reply:
Sorry, I just found out where I was wrong, it should be
```python
n_phi.assign(project(normalphi,VectorSpace1))
```
and I have checked that the results of these two methods are the same